File: EnhancedBulkCompressor.py
"""
Bulk Decompress Files on Cloud Storage

This template is a batch pipeline that decompresses files on Cloud Storage to a specified location.
It automatically handles multiple compression modes during a single run and determines the
decompression mode to use based on the file extension (.bzip2, .deflate, .gz).

Template Metadata:
    Name: Bulk_Decompress_GCS_Files
    Category: UTILITIES
    Description: Decompresses files from Cloud Storage to a specified location.
    Documentation: https://cloud.google.com/dataflow/docs/guides/templates/provided/bulk-decompress-cloud-storage

Note: This template is intended for single compressed files and not compressed folders.
"""

import apache_beam as beam
from apache_beam.options.pipeline_options import PipelineOptions
from apache_beam.io.filesystem import CompressionTypes
from apache_beam.io.gcp.gcsio import GcsIO
from google.cloud import storage
import logging
import csv
import io
import os
import multiprocessing
from concurrent.futures import ThreadPoolExecutor
from functools import partial
import time

# Set up logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Constants
SUPPORTED_COMPRESSIONS = {CompressionTypes.BZIP2, CompressionTypes.GZIP, CompressionTypes.DEFLATE}
UNCOMPRESSED_ERROR_MSG = "Skipping file %s because it did not match any compression mode (%s)"
MALFORMED_ERROR_MSG = "The file resource %s is malformed or not in %s compressed format."

class BulkDecompressorOptions(PipelineOptions):
    @classmethod
    def _add_argparse_args(cls, parser):
        parser.add_value_provider_argument(
            '--input_file_pattern',
            help='Input Cloud Storage file pattern (e.g., gs://your-bucket/compressed/*.gz)')
        parser.add_value_provider_argument(
            '--output_bucket',
            help='Output Cloud Storage bucket for decompressed files')
        parser.add_value_provider_argument(
            '--output_failure_file',
            help='Output file for decompression failures')
        parser.add_value_provider_argument(
            '--batch_size',
            type=int,
            default=100,
            help='Number of files to process in each batch')

    def validate(self, validator):
        validator.check_option_not_none(self, 'input_file_pattern')
        validator.check_option_not_none(self, 'output_bucket')
        validator.check_option_not_none(self, 'output_failure_file')

class Decompress(beam.DoFn):
    def __init__(self, output_bucket):
        self.output_bucket = output_bucket
        self.gcs_client = None
        self.pool = None

    def setup(self):
        self.gcs_client = storage.Client()
        self.pool = ThreadPoolExecutor(max_workers=multiprocessing.cpu_count())
        logger.info(f"GCS client and ThreadPoolExecutor initialized with {multiprocessing.cpu_count()} workers")

    def teardown(self):
        if self.pool:
            self.pool.shutdown()
        logger.info("ThreadPoolExecutor shut down")

    def process(self, element):
        start_time = time.time()
        for file_metadata in element:
            input_path = file_metadata.path
            output_path = input_path.replace(
                file_metadata.metadata.bucket, self.output_bucket.get()
            ).rsplit('.', 1)[0]

            logger.info(f"Processing file: {input_path}")

            if self._output_file_exists(output_path):
                logger.info(f"Output file already exists, skipping: {output_path}")
                yield output_path
                continue

            compression = CompressionTypes.detect_compression_type(input_path)
            if compression == CompressionTypes.UNCOMPRESSED:
                error_msg = UNCOMPRESSED_ERROR_MSG % (input_path, SUPPORTED_COMPRESSIONS)
                logger.warning(error_msg)
                yield beam.pvalue.TaggedOutput('errors', (input_path, error_msg))
                continue

            try:
                logger.info(f"Decompressing file: {input_path} with compression type: {compression}")
                future = self.pool.submit(self._decompress_and_write, input_path, output_path, compression)
                result = future.result()  # This will raise an exception if the decompression failed
                logger.info(f"Successfully decompressed: {input_path} to {output_path}")
                yield output_path
            except Exception as e:
                error_msg = self._sanitize_error_msg(str(e), input_path, compression)
                logger.error(f"Error decompressing {input_path}: {error_msg}")
                yield beam.pvalue.TaggedOutput('errors', (input_path, error_msg))

        end_time = time.time()
        logger.info(f"Batch processing time: {end_time - start_time:.2f} seconds")

    def _decompress_and_write(self, input_path, output_path, compression):
        start_time = time.time()
        with GcsIO().open(input_path, 'rb') as input_file:
            with GcsIO().open(output_path, 'wb') as output_file:
                decompressor = CompressionTypes.codec_for_compression_type(compression).get_decoder()
                while True:
                    chunk = input_file.read(1024 * 1024)  # Read 1MB at a time
                    if not chunk:
                        break
                    decompressed_chunk = decompressor.decompress(chunk)
                    output_file.write(decompressed_chunk)
                decompressed_chunk = decompressor.flush()
                if decompressed_chunk:
                    output_file.write(decompressed_chunk)
        end_time = time.time()
        logger.info(f"File decompression time for {input_path}: {end_time - start_time:.2f} seconds")

    def _output_file_exists(self, path):
        bucket_name, blob_name = path.replace('gs://', '').split('/', 1)
        bucket = self.gcs_client.bucket(bucket_name)
        exists = bucket.blob(blob_name).exists()
        logger.debug(f"Checked existence of {path}: {'Exists' if exists else 'Does not exist'}")
        return exists

    def _sanitize_error_msg(self, error_msg, input_file, compression):
        if "not in the BZip2 format" in error_msg or "incorrect header check" in error_msg:
            return MALFORMED_ERROR_MSG % (input_file, compression)
        return error_msg

class BatchElements(beam.DoFn):
    def __init__(self, batch_size):
        self.batch_size = batch_size

    def process(self, element):
        batch = []
        for item in element:
            batch.append(item)
            if len(batch) >= self.batch_size:
                yield batch
                batch = []
        if batch:
            yield batch

def run(argv=None):
    pipeline_options = BulkDecompressorOptions(argv)
    pipeline_options.view_as(beam.options.StandardOptions).runner = 'DataflowRunner'

    # Validate options
    pipeline_options.validate(beam.options.pipeline_options.PipelineOptionsValidator())

    logger.info("Starting Bulk Decompressor pipeline")
    with beam.Pipeline(options=pipeline_options) as p:
        matched_files = (p 
            | 'MatchFiles' >> beam.io.fileio.MatchFiles(pipeline_options.input_file_pattern)
            | 'BatchFiles' >> beam.ParDo(BatchElements(pipeline_options.batch_size)))

        logger.info("Applying Decompress DoFn")
        decompressed_files = (matched_files
            | 'Decompress' >> beam.ParDo(Decompress(pipeline_options.output_bucket))
            .with_outputs('errors', main='decompressed'))

        # Write successful decompressions
        decompressed_files.decompressed | 'LogSuccess' >> beam.Map(logger.info)

        # Format and write errors
        (decompressed_files.errors
            | 'FormatErrors' >> beam.Map(lambda x: f'{x[0]},{x[1]}')
            | 'WriteErrors' >> beam.io.WriteToText(
                pipeline_options.output_failure_file,
                header='Filename,Error',
                shard_name_template=''))

    logger.info("Bulk Decompressor pipeline completed")

if __name__ == '__main__':
    logging.getLogger().setLevel(logging.INFO)
    run()

File: EnhancedBulkCompressor.java
// Import necessary libraries and packages
package com.google.cloud.teleport.templates;

import com.google.cloud.teleport.metadata.Template;
import com.google.cloud.teleport.metadata.TemplateCategory;
import com.google.cloud.teleport.metadata.TemplateParameter;
import com.google.cloud.teleport.templates.BulkDecompressor.Options;
import com.google.common.annotations.VisibleForTesting;
import com.google.common.collect.ImmutableList;
import com.google.common.io.ByteStreams;
import com.google.common.io.Files;
import java.io.IOException;
import java.io.StringWriter;
import java.nio.channels.ReadableByteChannel;
import java.nio.channels.WritableByteChannel;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import javax.annotation.Nullable;
import org.apache.beam.sdk.Pipeline;
import org.apache.beam.sdk.PipelineResult;
import org.apache.beam.sdk.io.Compression;
import org.apache.beam.sdk.io.FileIO;
import org.apache.beam.sdk.io.FileSystems;
import org.apache.beam.sdk.io.TextIO;
import org.apache.beam.sdk.io.fs.MatchResult;
import org.apache.beam.sdk.io.fs.MoveOptions;
import org.apache.beam.sdk.io.fs.ResolveOptions.StandardResolveOptions;
import org.apache.beam.sdk.io.fs.ResourceId;
import org.apache.beam.sdk.options.PipelineOptions;
import org.apache.beam.sdk.options.PipelineOptionsFactory;
import org.apache.beam.sdk.options.Validation.Required;
import org.apache.beam.sdk.options.ValueProvider;
import org.apache.beam.sdk.transforms.DoFn;
import org.apache.beam.sdk.transforms.MapElements;
import org.apache.beam.sdk.transforms.ParDo;
import org.apache.beam.sdk.util.MimeTypes;
import org.apache.beam.sdk.values.KV;
import org.apache.beam.sdk.values.PCollectionTuple;
import org.apache.beam.sdk.values.TupleTag;
import org.apache.beam.sdk.values.TupleTagList;
import org.apache.beam.sdk.values.TypeDescriptors;
import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVPrinter;
import org.apache.commons.csv.QuoteMode;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

// Define the template metadata
@Template(
    name = "Bulk_Decompress_GCS_Files",
    category = TemplateCategory.UTILITIES,
    displayName = "Bulk Decompress Files on Cloud Storage",
    description = {
      "The Bulk Decompress Cloud Storage Files template is a batch pipeline that decompresses files on Cloud Storage to a specified location. "
          + "This functionality is useful when you want to use compressed data to minimize network bandwidth costs during a migration, but would like to maximize analytical processing speed by operating on uncompressed data after migration. "
          + "The pipeline automatically handles multiple compression modes during a single run and determines the decompression mode to use based on the file extension (.bzip2, .deflate, .gz, .zip).",
      "Note: The Bulk Decompress Cloud Storage Files template is intended for single compressed files and not compressed folders."
    },
    optionsClass = Options.class,
    documentation =
        "https://cloud.google.com/dataflow/docs/guides/templates/provided/bulk-decompress-cloud-storage",
    contactInformation = "https://cloud.google.com/support",
    requirements = {
      "The files to decompress must be in one of the following formats: Bzip2, Deflate, and Gzip.",
      "The output bucket must exist prior to running the pipeline."
    })
public class BulkDecompressor {

  // Logger for this class
  private static final Logger LOG = LoggerFactory.getLogger(BulkDecompressor.class);

  // Define supported compression types
  @VisibleForTesting
  static final Set<Compression> SUPPORTED_COMPRESSIONS =
      Stream.of(Compression.values())
          .filter(value -> value != Compression.AUTO && value != Compression.UNCOMPRESSED)
          .collect(Collectors.toSet());

  // Error message for uncompressed files
  @VisibleForTesting
  static final String UNCOMPRESSED_ERROR_MSG =
      "Skipping file %s because it did not match any compression mode (%s)";

  // Error message for malformed compressed files
  @VisibleForTesting
  static final String MALFORMED_ERROR_MSG =
      "The file resource %s is malformed or not in %s compressed format.";

  // Define output tags for the main decompression output and errors
  @VisibleForTesting
  static final TupleTag<String> DECOMPRESS_MAIN_OUT_TAG = new TupleTag<String>() {};

  @VisibleForTesting
  static final TupleTag<KV<String, String>> DEADLETTER_TAG = new TupleTag<KV<String, String>>() {};

  // Define pipeline options interface
  public interface Options extends PipelineOptions {
    @TemplateParameter.GcsReadFile(
        order = 1,
        groupName = "Source",
        description = "Input Cloud Storage File(s)",
        helpText = "The Cloud Storage location of the files you'd like to process.",
        example = "gs://your-bucket/your-files/*.gz")
    @Required
    ValueProvider<String> getInputFilePattern();

    void setInputFilePattern(ValueProvider<String> value);

    @TemplateParameter.GcsWriteFolder(
        order = 2,
        groupName = "Target",
        description = "Output bucket in Cloud Storage",
        helpText = "The Cloud Storage bucket where decompressed files will be written, maintaining the original directory structure.",
        example = "gs://your-output-bucket")
    @Required
    ValueProvider<String> getOutputBucket();

    void setOutputBucket(ValueProvider<String> value);

    @TemplateParameter.GcsWriteFile(
        order = 3,
        description = "The output file for failures during the decompression process",
        helpText =
            "The output file to write failures to during the decompression process. If there are no failures, the file will still be created but will be empty. The contents will be one line for each file which failed decompression in CSV format (Filename, Error). Note that this parameter will allow the pipeline to continue processing in the event of a failure.",
        example = "gs://your-bucket/decompressed/failed.csv")
    @Required
    ValueProvider<String> getOutputFailureFile();

    void setOutputFailureFile(ValueProvider<String> value);
  }

  // Main method to run the pipeline
  public static void main(String[] args) {
    Options options = PipelineOptionsFactory.fromArgs(args).withValidation().as(Options.class);
    run(options);
  }

  // Method to set up and run the pipeline
  public static PipelineResult run(Options options) {
    Pipeline pipeline = Pipeline.create(options);

    // Define the main decompression logic
    PCollectionTuple decompressOut =
        pipeline
            .apply("MatchFile(s)", FileIO.match().filepattern(options.getInputFilePattern()))
            .apply(
                "DecompressFile(s)",
                ParDo.of(new Decompress(options.getOutputBucket()))
                    .withOutputTags(DECOMPRESS_MAIN_OUT_TAG, TupleTagList.of(DEADLETTER_TAG)));

    // Handle and format errors
    decompressOut
        .get(DEADLETTER_TAG)
        .apply(
            "FormatErrors",
            MapElements.into(TypeDescriptors.strings())
                .via(
                    kv -> {
                      StringWriter stringWriter = new StringWriter();
                      try {
                        CSVPrinter printer =
                            new CSVPrinter(
                                stringWriter,
                                CSVFormat.DEFAULT
                                    .withEscape('\\')
                                    .withQuoteMode(QuoteMode.NONE)
                                    .withRecordSeparator('\n'));
                        printer.printRecord(kv.getKey(), kv.getValue());
                      } catch (IOException e) {
                        throw new RuntimeException(e);
                      }

                      return stringWriter.toString();
                    }))
        .apply(
            "WriteErrorFile",
            TextIO.write()
                .to(options.getOutputFailureFile())
                .withHeader("Filename,Error")
                .withoutSharding());

    return pipeline.run();
  }

  // DoFn class for decompression logic
  @SuppressWarnings("serial")
  public static class Decompress extends DoFn<MatchResult.Metadata, String> {
    private final ValueProvider<String> outputBucket;

    // Constructor
    Decompress(ValueProvider<String> outputBucket) {
      this.outputBucket = outputBucket;
    }

    // Main processing method
    @ProcessElement
    public void processElement(ProcessContext context) {
      ResourceId inputFile = context.element().resourceId();
      ResourceId outputFile = getOutputFile(inputFile);

      // Check if output file already exists
      if (outputFileExists(outputFile)) {
        context.output(outputFile.toString());
        return;
      }

      // Check if input file is compressed
      if (!Compression.AUTO.isCompressed(inputFile.toString())) {
        String errorMsg = String.format(UNCOMPRESSED_ERROR_MSG, inputFile.toString(), SUPPORTED_COMPRESSIONS);
        context.output(DEADLETTER_TAG, KV.of(inputFile.toString(), errorMsg));
        return;
      }

      // Attempt decompression
      try {
        decompress(inputFile, outputFile);
        context.output(outputFile.toString());
      } catch (IOException e) {
        LOG.error(e.getMessage());
        context.output(DEADLETTER_TAG, KV.of(inputFile.toString(), e.getMessage()));
      }
    }

    // Helper method to get the output file path
    private ResourceId getOutputFile(ResourceId inputFile) {
      String inputPath = inputFile.toString();
      String outputPath = inputPath.replace(inputFile.getBucket(), outputBucket.get());
      return FileSystems.matchNewResource(Files.getNameWithoutExtension(outputPath), false);
    }

    // Helper method to check if output file exists
    private boolean outputFileExists(ResourceId outputFile) {
      try {
        FileSystems.match(outputFile.toString());
        return true;
      } catch (IOException e) {
        return false;
      }
    }

    // Helper method to perform decompression
    private void decompress(ResourceId inputFile, ResourceId outputFile) throws IOException {
      Compression compression = Compression.detect(inputFile.toString());
      
      try (ReadableByteChannel readerChannel = compression.readDecompressed(FileSystems.open(inputFile))) {
        try (WritableByteChannel writerChannel = FileSystems.create(outputFile, MimeTypes.TEXT)) {
          ByteStreams.copy(readerChannel, writerChannel);
        }
      } catch (IOException e) {
        String msg = e.getMessage();
        LOG.error("Error occurred during decompression of {}", inputFile.toString(), e);
        throw new IOException(sanitizeDecompressionErrorMsg(msg, inputFile, compression));
      }
    }

    // Helper method to sanitize decompression error messages
    private String sanitizeDecompressionErrorMsg(
        @Nullable String errorMsg, ResourceId inputFile, Compression compression) {
      if (errorMsg != null
          && (errorMsg.contains("not in the BZip2 format")
              || errorMsg.contains("incorrect header check"))) {
        errorMsg = String.format(MALFORMED_ERROR_MSG, inputFile.toString(), compression);
      }
      return errorMsg == null ? "" : errorMsg;
    }
  }
}
